### PR TITLE
Enable analyzers for mono/netcore

### PR DESCRIPTION
**Pull request by stephentoub**
cc: @akoeplinger, @marek-safar

This is mostly just running auto-fixers for the various violated rules.  In just a few places, I manually intervened.

### CODE REVIEW

File: src/mono/netcore/System.Private.CoreLib/src/System/Reflection/Emit/AssemblyBuilder.Mono.cs [hunk 1]
```diff
@@ -48,544 +47,598 @@
...
+            SetCustomAttribute(new CustomAttributeBuilder(con, binaryAttribute));
+        }
+
+        private Exception not_supported()
+        {
+            // Strange message but this is what MS.NET prints...
+            return new NotSupportedException("The invoked member is not supported in a dynamic module.");
+        }
+
+        private string create_assembly_version(string version)
```
Comment (danmosemsft) [inline src/mono/netcore/System.Private.CoreLib/src/System/Reflection/Emit/AssemblyBuilder.Mono.cs]: Nit, Some method naming could be improved later

File: src/mono/netcore/System.Private.CoreLib/src/System/Reflection/Emit/ConstructorOnTypeBuilderInst.cs [hunk 2]
```diff
@@ -36,192 +36,225 @@
...
+            {
+                return instantiation;
+            }
+        }
+
+        public override string Name
+        {
+            get
+            {
+                return cb.Name;
```
Comment (danmosemsft) [inline src/mono/netcore/System.Private.CoreLib/src/System/Reflection/Emit/ConstructorOnTypeBuilderInst.cs]: There's a lot of simple property getters that could become expression bodied. Are there fixers to do that?

Comment (stephentoub) [reply src/mono/netcore/System.Private.CoreLib/src/System/Reflection/Emit/ConstructorOnTypeBuilderInst.cs]: Yes, but I only fixed things for analyzers that we have set to produce warnings.

File: src/mono/netcore/System.Private.CoreLib/src/System/Reflection/Emit/TypeBuilder.Mono.cs [hunk 3]
```diff
@@ -47,1846 +46,2030 @@
...
+                    if (t == interfaceType)
+                        return;
+
+                Type[] ifnew = new Type[interfaces.Length + 1];
+                interfaces.CopyTo(ifnew, 0);
+                ifnew[interfaces.Length] = interfaceType;
+                interfaces = ifnew;
+            }
+            else
+
```
Comment (danmosemsft) [inline src/mono/netcore/System.Private.CoreLib/src/System/Reflection/Emit/TypeBuilder.Mono.cs]: SR.NotSupported_NoParentDefaultConstructor ..

File: src/mono/netcore/System.Private.CoreLib/src/System/Reflection/Emit/TypeBuilder.Mono.cs [hunk 4]
```diff
@@ -47,1846 +46,2030 @@
...
-				throw new ArgumentException ("type is not TypeBuilder but " + type.GetType (), "type");
-
-			if (type is TypeBuilder && type.ContainsGenericParameters)
-				type = type.MakeGenericType (type.GetGenericArguments ());
-
-			if (!type.IsGenericType)
-				throw new ArgumentException ("type is not a generic type", "type");
-
-			if (!method.DeclaringType.IsGenericTypeDefinition)
-				throw new ArgumentException ("method declaring type is not a generic type definition", "method");
```
Comment (danmosemsft) [inline src/mono/netcore/System.Private.CoreLib/src/System/Reflection/Emit/TypeBuilder.Mono.cs]: do we have an issue already to move strings into resources? or in most cases, find the existing resources

File: src/mono/netcore/System.Private.CoreLib/src/System/TypedReference.cs [hunk 5]
```diff
@@ -7,97 +7,103 @@
...
-		public unsafe static void SetTypedReference (TypedReference target, object? value)
-		{
-			throw new NotSupportedException ();
-		}
-	}
+    [CLSCompliantAttribute(false)]
+    public ref struct TypedReference
+    {
+        #region sync with object-internals.h
+        #pragma warning disable CA1823
```
Comment (danmosemsft) [inline src/mono/netcore/System.Private.CoreLib/src/System/TypedReference.cs]: `// used by runtime` I assume?

Comment (stephentoub) [reply src/mono/netcore/System.Private.CoreLib/src/System/TypedReference.cs]: I figured that was obvious from the comment/region right above it, but I can add it explicitly if it's be helpful.

**Review by danmoseley — APPROVED** [review]

File: src/mono/netcore/System.Private.CoreLib/src/System/TypedReference.cs (hunk 5 above)
Comment (danmosemsft) [reply src/mono/netcore/System.Private.CoreLib/src/System/TypedReference.cs]: fair enough

File: src/mono/netcore/System.Private.CoreLib/src/System/Reflection/Emit/TypeBuilder.Mono.cs (hunk 4 above)
Comment (stephentoub) [reply src/mono/netcore/System.Private.CoreLib/src/System/Reflection/Emit/TypeBuilder.Mono.cs]: Not sure.  It'd be good to do, though.  There are probably several passes that should be made through the mono-specific corelib sources.  This PR is just enough to turn on our analyzers.

Comment (danmosemsft) [reply src/mono/netcore/System.Private.CoreLib/src/System/Reflection/Emit/TypeBuilder.Mono.cs]: I'll open one

Comment (danmosemsft) [reply src/mono/netcore/System.Private.CoreLib/src/System/Reflection/Emit/TypeBuilder.Mono.cs]: https://github.com/dotnet/runtime/issues/34056

**Comment by danmosemsft** [comment]
```
src\libraries\Common\src\Interop\Windows\User32\Interop.GetProcessWindowStation.cs(12,9): error BCL0015: (NETCORE_ENGINEERING_TELEMETRY=Build) user32.dll!GetProcessWindowStation is not supported on one\more targeted platforms
```
I get the same locally.

I pushed a change to disable it pending https://github.com/dotnet/runtime/issues/33867

**Review by marek-safar — APPROVED** [review]

File: src/mono/netcore/System.Private.CoreLib/System.Private.CoreLib.csproj [hunk 6]
```diff
@@ -9,6 +9,7 @@
     <GenerateNeutralResourcesLanguageAttribute>false</GenerateNeutralResourcesLanguageAttribute>
     <EnsureRuntimePackageDependencies>false</EnsureRuntimePackageDependencies>
     <TargetFramework>netcoreapp2.1</TargetFramework>
+    <EnablePInvokeAnalyzer>false</EnablePInvokeAnalyzer>
```
Comment (akoeplinger) [inline src/mono/netcore/System.Private.CoreLib/System.Private.CoreLib.csproj]: should this be reenabled at some point? if yes we should add a TODO or file an issue.

**Review by akoeplinger — APPROVED** [review]
LGTM with one question.

File: src/mono/netcore/System.Private.CoreLib/System.Private.CoreLib.csproj (hunk 6 above)
Comment (marek-safar) [reply src/mono/netcore/System.Private.CoreLib/System.Private.CoreLib.csproj]: There is already ##33867

**Comment by akoeplinger** [comment]
The test failure is the known issue: https://github.com/dotnet/runtime/issues/34045

**Comment by akoeplinger** [comment]
@stephentoub by the way how did you run these auto-fixers?

**Comment by stephentoub** [comment]
> @stephentoub by the way how did you run these auto-fixers?

Opened the solution in VS, and used Build | Run Code Analysis on Solution to force all of the analyzers to run promptly.  Then for each category, selected one of the warnings from the error list, triggered its fixer from the lightbulb (ctrl-period), and selected Solution when determining the scope to which to apply it.  Repeated that for each kind of warning.  Then also hacked the .csproj with manually specifying TargetsLinux/OSX/Windows and repeated the process, to cover files only in specific builds.

On that note, how do I build mono's corelib for Linux and macOS?  I was able to specify a property on the command-line, and it output the assembly into the relevant location, but it appeared to still be building the Windows source code.

**Comment by akoeplinger** [comment]
> On that note, how do I build mono's corelib for Linux and macOS? I was able to specify a property on the command-line, and it output the assembly into the relevant location, but it appeared to still be building the Windows source code.

As far as I know everything should use the `TargetOS` msbuild property, same as CoreCLR. What property did you pass?

**Comment by stephentoub** [comment]
> As far as I know everything should use the TargetOS msbuild property, same as CoreCLR. What property did you pass?

That appears to work.  No idea what I was doing differently before :)

**Comment by danmosemsft** [comment]
> triggered its fixer from the lightbulb (ctrl-period), and selected Solution when determining the scope to which to apply it. Repeated that for each kind of warning.

From a code review point of view, consider making a commit after each fixer. That way I can do only brief review of eg "remove unused usings" and focus on other stuff.

**Pull request merged**